libmisc/strscpy.c

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2022, Alejandro Colomar <alx.manpages@gmail.com>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the copyright holders or contributors may not be used to
+ *    endorse or promote products derived from this software without
+ *    specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+ * PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stddef.h>
+#include <sys/types.h>
+#include <string.h>
+
+#ident "$Id$"
+
+#include "defines.h"
+#include "prototypes.h"
+
+
+/*
+ * ARGUMENTS
+ *	dst	Pointer to the first byte of the destination buffer.
+ *
+ *	src	Pointer to the first byte of the source string.
+ *
+ *	n	Size of the 'dst' and 'src' buffers.
+ *
+ * DESCRIPTION
+ *	Copy a string that may not be NUL-terminated into a buffer.
+ *	If the original string is longer than the buffer, copy as much
+ *	of it as fits.  The resulting string is guaranteed to be
+ *	NUL-terminated, if the buffer size is at least 1.
+ *
+ * RETURN VALUE
+ *	On success, this function returns the length of the resulting
+ *	string.
+ *
+ *	If the input string didn't contain a NUL character, -E2BIG is
+ *	returned.
+ *
+ * CAVEATS
+ *	This function can hide bugs in the input string, by not reading
+ *	until the end of the string.  strlcpy(3bsd) and stpecpy()
+ *	read the input string until the end of it, which normally
+ *	results in SIGSEGV if the input string was not correctly
+ *	NUL-terminated.
```

Comment:
Should also add the caveat that results will be undefined if memory areas overlap (per memccpy manpage)

Reply:
That should be obvious by the use of `restrict` in the prototype.  I also updated the Linux man-pages recently to document restrict in the SYNOPSIS sections consistently.  I think the only function that explicitly states in text that overlapping memory is UB is memcpy(3), due to the historical bugs about it.

I mean, we could add the text, but my first impression is that I prefer to keep it short.  `restrict` is clear (to those who know the keyword, that is).